github/ratelimit_test.go

```diff
@@ -0,0 +1,331 @@
+	url := req.URL.String()
+	tc, ok := testCases[url]
+	if !ok {
+		return newResponse(apiResponse{}, nil, http.StatusOK)
+	}
+
+	if tc.err != nil {
+		return &http.Response{}, tc.err
+	}
+
+	return newResponse(tc.response, tc.headers, tc.statusCode)
+}
+
+func TestRateLimitSuite(t *testing.T) {
+	suite.Run(t, new(RateLimitSuite))
+}
+
+type RateLimitSuite struct {
+	suite.Suite
+	transport  *RateLimitTransport
+	loggerMock *testutils.LoggerMock
+}
+
+func (s *RateLimitSuite) SetupSuite() {
+	defaultAbuseRetryAfter = default403Reset
+}
+
+func (s *RateLimitSuite) SetupTest() {
+	s.loggerMock = &testutils.LoggerMock{}
+	s.transport = NewRateLimitTransport(&gitHubTransportResponseMock{}, s.loggerMock)
+}
+
+func (s *RateLimitSuite) TearDownSuite() {
+}
+
+// TestNoLimit ensures that the regular case: no rate limit nor abuse, is not considered a RateLimit case
+func (s *RateLimitSuite) TestNoLimit() {
+	t0 := time.Now()
+	response, err := s.transport.RoundTrip(newRequest("/normal"))
+	elapsed := time.Now().Sub(t0)
+	s.True(elapsed < 500*time.Millisecond, "request took %s, but it should be almost instant", elapsed)
+	s.Equal("", s.loggerMock.Next())
+	s.NoError(err)
+
+	content, err := ioutil.ReadAll(response.Body)
+	s.NoError(err)
+	var data apiResponse
+	err = json.Unmarshal(content, &data)
+	s.NoError(err)
+	s.Equal("success", data.Data)
+	s.Len(data.Errors, 0)
+}
+
+// TestRateLimitConsecutively ensures that hitting RateLimit twice, causes two waiting periods
+// it should log the RateLimitError, and wait till the lock expires
+// if the next Request hits the RateLimit again, it should do the same
+// if the next Request does not hit the RateLimit, it should return the response inmediately
+func (s *RateLimitSuite) TestRateLimitConsecutively() {
+	t0 := time.Now()
+	response, err := s.transport.RoundTrip(newRequest("/ratelimit_sleep"))
+	elapsed := time.Now().Sub(t0)
+	s.True(elapsed < 500*time.Millisecond, "request took %s, but it should be almost instant", elapsed)
+	s.Regexp("locking transport for [^;]+; API rate limit exceeded", s.loggerMock.Next())
+	s.Equal("", s.loggerMock.Next())
+	s.IsType(&ErrRateLimit{}, err)
+	s.Nil(response)
+
+	response, err = s.transport.RoundTrip(newRequest("/ratelimit_sleep"))
+	t1 := time.Now()
+	elapsed = t1.Sub(t0)
+	s.True(elapsed > defaultRateLimitReset, "request took %s, but it should be, at least %s", elapsed, defaultRateLimitReset)
+	s.Contains(s.loggerMock.Next(), "rate limit reached, sleeping until")
+	s.Regexp("locking transport for [^;]+; API rate limit exceeded", s.loggerMock.Next())
+	s.Equal("", s.loggerMock.Next())
+	s.IsType(&ErrRateLimit{}, err)
+	s.Nil(response)
+
+	response, err = s.transport.RoundTrip(newRequest("/normal"))
+	t2 := time.Now()
+	elapsed = t2.Sub(t1)
+	s.True(elapsed > defaultRateLimitReset, "request took %s, but it should be, at least %s", elapsed, defaultRateLimitReset)
+	s.Contains(s.loggerMock.Next(), "rate limit reached, sleeping until")
```

Comment:
I'm a little confused about this logic.
Aren't transport sleep when the limit is reached and locks any other requests? And executes next request only when limit was reset? Then why this request logs sleeping 🤔 

I mean your code:
```go
	rt.Lock()
	defer rt.Unlock()

	if time.Now().Before(rt.lockedUntil) {
		rt.logger.Infof("rate limit reached, sleeping until %s", rt.lockedUntil)
		time.Sleep(rt.lockedUntil.Sub(time.Now()))
	}
```

logging shouldn't be executed unless the round tripper was unlocked. And it's unlocked only after reset.

Reply:
I don't understand your questions.
The `RateLimitTransport` only sleeps when it should avoid a new call to GH, not before; example:
- a request triggers an `ErrAbuse` -> enables the lock (instead of sleep) -> returns `ErrAbuse`
- it comes a new request:
  - the lock already expired -> process it.
  - the lock didn't expire yet -> sleep to avoid calling GH.

Why?
because returning the `ErrAbuse` instead of sleeping will let the caller to decide what to do:
- retry, (what we're doing right now)
- discard,
- change token...

Reply:
Yeah. But in the code I see 3 calls to the API. According to your logic:
1. Line 190: a request triggers an ErrAbuse -> enables the lock (instead of sleep) -> returns ErrAbuse
2. Line 198: the lock didn't expire yet -> sleep to avoid calling GH.
3. Line 208: the lock already expired (because of the sleep above) -> process it.

What am I missing?

Reply:
Ah. The second call returns error again so this request sleeps. Got it. Thanks!